### PR TITLE
[VCARB-121] - Start using the config.toml

### PR DESCRIPTION
**Pull request by potofpie**
- Read the token file if we dont have a parameter
- Create datadir correctly if we dont pass a parameter
-  Add the serverId in the POST to create the session

### CODE REVIEW

File: cmd/enroll.go [hunk 1]
```diff
@@ -107,5 +107,5 @@ func init() {
 	rootCmd.AddCommand(enrollCmd)
 	enrollCmd.Flags().String("api-url", "", "the for testing again preview environment")
 	enrollCmd.Flags().MarkHidden("api-url")
-	enrollCmd.Flags().String("data-dir", cwd, "the data directory for storing logs and other data")
+	enrollCmd.Flags().String("data-dir", filepath.Join(cwd+"/datadir"), "the data directory for storing logs and other data")
```
Comment (jhaynie) [inline cmd/enroll.go:110]:
```suggestion
	enrollCmd.Flags().String("data-dir", filepath.Join(cwd, "datadir"), "the data directory for storing logs and other data")
```

File: cmd/root.go [hunk 2]
```diff
@@ -37,6 +38,28 @@ func mustFlagString(cmd *cobra.Command, name string, required bool) string {
 	return val
 }
 
+type EnrollTokenData struct {
```
Comment (jhaynie) [inline cmd/root.go:41]: we should use viper instead. viper is built by the same people that built cobra and it makes it trivial to create configs that also bind to command line arguments and environment variables.

https://github.com/spf13/viper

also we should probably choose something a bit more user friendly in the config format like TOML

**Review by jhaynie — CHANGES_REQUESTED** [review]

File: cmd/enroll.go [hunk 3]
```diff
@@ -1,20 +1,24 @@
...
+	"bytes"
 	"encoding/json"
 	"errors"
 	"fmt"
+	"log"
 	"net/http"
 	"os"
 	"path/filepath"
 
+	"github.com/BurntSushi/toml"
```
Comment (robindiddams) [inline cmd/enroll.go:13]: yoooo BurntSushi! that takes me back!

File: cmd/enroll.go [hunk 4]
```diff
@@ -1,20 +1,24 @@
...
+
 	"github.com/shopmonkeyus/eds-server/internal/util"
 	"github.com/spf13/cobra"
 )
 
 type enrollTokenData struct {
-	Token    string `json:"token"`
-	ServerID string `json:"serverId"`
+	Token    string `json:"token" toml:"server_id`
```
Comment (robindiddams) [inline cmd/enroll.go:20]:
```suggestion
	Token    string `json:"token" toml:"token`
```
ya gotta get that golang vscode plugin, `go vet ./...` will catch this for you

File: cmd/server.go [hunk 5]
```diff
@@ -894,6 +913,7 @@ var serverHelpCmd = &cobra.Command{
 }
 
 func init() {
+	cobra.OnInitialize(initConfig)
```
Comment (robindiddams) [inline cmd/server.go:916]: this should go in root.go

File: cmd/fork.go [hunk 6]
```diff
@@ -281,11 +282,16 @@ var forkCmd = &cobra.Command{
...
+	if err != nil {
+		fmt.Println("couldn't get current working directory: ", err)
+		os.Exit(1)
+	}
 	rootCmd.AddCommand(forkCmd)
 	forkCmd.Hidden = true // don't expose this since its only called by the main server process in the wrapper
 
 	// NOTE: sync these with serverCmd
-	forkCmd.Flags().String("data-dir", "", "the data directory for storing state, logs, and other data")
+	forkCmd.Flags().String("data-dir", filepath.Join(cwd, "dataDir"), "the data directory for storing state, logs, and other data")
```
Comment (robindiddams) [inline cmd/fork.go:294]: lets just make data-dir a persistent flag on the rootCmd

File: cmd/root.go [hunk 7]
```diff
@@ -39,6 +41,20 @@ func mustFlagString(cmd *cobra.Command, name string, required bool) string {
...
+	Token    string `json:"token"`
+	ServerID string `json:"serverId"`
+}
+
+func initConfig() {
+	viper.SetConfigName("config")
+	viper.AddConfigPath("./dataDir")
```
Comment (robindiddams) [inline cmd/root.go:51]: you cant use `/` in path names this breaks on windows. also you're hard coding this here but allowing it to be passed in elswhere which will conflict

File: cmd/server.go [hunk 8]
```diff
@@ -546,15 +549,31 @@ var serverCmd = &cobra.Command{
...
+		apiurl := mustFlagString(cmd, "api-url", false)
 		driverURL := mustFlagString(cmd, "url", true)
-		server := mustFlagString(cmd, "server", true)
+		server := mustFlagString(cmd, "server", false)
+		edsServerId := mustFlagString(cmd, "eds-server-id", false)
 		dataDir := getDataDir(cmd, logger)
+		apikey := mustFlagString(cmd, "api-key", false)
+
+		if apikey == "" {
+			apikey = viper.GetString("token")
```
Comment (robindiddams) [inline cmd/server.go:560]: should just bind the viper flag `token` to the flag `api-key`

**Review by robindiddams — CHANGES_REQUESTED** [review]
Enroll fails if theres no config file yet, should probably check the config file differently in that case. Also theres some weirdness with the viper config, [heres an example](https://gist.github.com/robindiddams/bb7db28ba9efc1b9517e18aaf10349f6) setup that lets you sorta configure it globally. Also that flag

`go mod tidy` to cleanup that mod file

File: cmd/root.go [hunk 9]
```diff
@@ -39,6 +42,18 @@ func mustFlagString(cmd *cobra.Command, name string, required bool) string {
...
+type EnrollTokenData struct {
+	Token    string `json:"token"`
+	ServerID string `json:"serverId"`
+}
+
+func initConfig() {
+	if cfgFile != "" {
+		// Use config file from the flag.
+		viper.SetConfigFile(cfgFile)
+	}
```
Comment (robindiddams) [inline cmd/root.go:54]: still need 

```go
	if err := viper.ReadInConfig(); err != nil {
   // 🤷‍♀️
	}
```

File: cmd/root.go [hunk 10]
```diff
@@ -239,10 +254,19 @@ func Execute() {
...
+	cobra.OnInitialize(initConfig)
 	rootCmd.PersistentFlags().Bool("verbose", false, "turn on verbose logging")
 	rootCmd.PersistentFlags().Bool("silent", false, "turn off all logging")
 	rootCmd.PersistentFlags().Bool("timestamp", false, "turn on timestamps in logs")
 	rootCmd.PersistentFlags().String("log-file-sink", "", "the log file sink to use")
 	rootCmd.PersistentFlags().MarkHidden("log-file-sink")
 	rootCmd.PersistentFlags().String("schema-validator", "", "the schema validator directory to use")
+	rootCmd.PersistentFlags().String("data-dir", filepath.Join(cwd, "dataDir"), "the data directory for storing state, logs, and other data")
+	rootCmd.PersistentFlags().String("config", filepath.Join(cwd, "dataDir/config.toml"), "the data directory for storing state, logs, and other data")
```
Comment (robindiddams) [inline cmd/root.go:270]: N/O S/L/A/S/H/E/S

```suggestion
	rootCmd.PersistentFlags().String("config", filepath.Join(cwd, "dataDir", "config.toml"), "the data directory for storing state, logs, and other data")
```

**Review by robindiddams — CHANGES_REQUESTED** [review]

File: cmd/root.go (hunk 10 above)
Comment (potofpie) [reply cmd/root.go:270]: lol true!

File: cmd/root.go [hunk 11]
```diff
@@ -239,10 +256,24 @@ func Execute() {
...
 func init() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		fmt.Println("couldn't get current working directory: ", err)
+		os.Exit(1)
+	}
+	if err != nil {
+		fmt.Printf("error getting current working directory: %s\n", err)
+		os.Exit(1)
+	}
```
Comment (robindiddams) [inline cmd/root.go:267]: extra error check?

File: cmd/enroll.go [hunk 12]
```diff
@@ -81,16 +84,19 @@ var enrollCmd = &cobra.Command{
 			logger.Fatal("failed to start enroll: %s", enrollResp.Message)
 		}
 
-		tokenFile := filepath.Join(dataDir, "token.json")
+		var buf bytes.Buffer
+		if err := toml.NewEncoder(&buf).Encode(enrollResp.Data); err != nil {
+			log.Fatal(err)
```
Comment (robindiddams) [inline cmd/enroll.go:89]: `logger` not `log` dont use the log package

**Review by robindiddams — CHANGES_REQUESTED** [review]
this isnt right, the full path is supposed to be the default. We should just have one flag for the data dir and `initConfig` should just append `config.toml` to the end of it.

**Review by robindiddams — APPROVED** [review]

**Pull request merged**